### PR TITLE
Implementation of extended DSHOT telemetry, smooth temperature limiti…

### PR DESCRIPTION
**Pull request by damosvil**
…ng function and wdt (currently disabled)

### CODE REVIEW

**Comment by stylesuxx** [comment]
Alright, to we have a lot of stuff going on in this PR:
1. Extended Dshot Telemetry
2. Watchdog implementation - disabled by default, can only be enabled via code
3. Improved handling of temperature

There are a lot of lines where only indentation has changed this pollutes the PR a little bit.
Looking at it, I am not sure if it wouldn't make sense to split it in 3 PRs so it is easier digestible...

I added some comments for the stuff that quickly caught my eye. I will get into the logic with a fresh mind tomorrow.

**Comment by saidinesh5** [comment]
Also we have to test watchdog implementation with the longest melody:
```
a,b,c,d,(255,255)*62
```
and make sure it doesn't cause boot loop because we disable all interrupts

**Comment by damosvil** [comment]
@saidinesh5 In principle WDT implementation is there but disabled by default. I have been playing with it for some time, but I found that at least 0.8ms is not enough for all motor types, so that sometimes it causes the esc to reboot, at least in the workbench. After keeping my drones armed for 5 minutes in my workbech, and being sure that no esc reboot I have flown them with no issues, but it takes extra time testing the drone. 
Please, let me know if you think it is better to remove all the code related to WDT from this pr.

**Comment by saidinesh5** [comment]
Yeah watchdog related code should be under a separate PR anyway because it isn't related to extended dshot anyway. Similarly smooth temperature limitation. 

Also, what about beacon melodies? ( The other PR that is currently pending).

**Comment by stylesuxx** [comment]
> Also, what about beacon melodies? ( The other PR that is currently pending).

EDT is currently the most important thing since those changes is also what the [documentation](https://github.com/bird-sanctuary/bluejay-documentation) is based on. Once EDT is merged, we are on par with documentation and can then merge the other stuff step by step.

I created [Milestones](https://github.com/bird-sanctuary/bluejay/milestones) - would just like to move the discussion to either Discord or the accompanying ticket to not pollute this PR.

**Comment by damosvil** [comment]
@saidinesh5 Agreed, so I removed all WDT related code.

**Comment by howels** [comment]
FYI betaflight merged extended dshot telemetry so now would be a good time to add a release or pre-release to esc-configurator for people to test with.

**Comment by stylesuxx** [comment]
> FYI betaflight merged extended dshot telemetry so now would be a good time to add a release or pre-release to esc-configurator for people to test with.

I will add a pre-release once we have things cleaned up here and finished internal testing.

**Review by stylesuxx — CHANGES_REQUESTED** [review]

File: Bluejay.asm [hunk 1]
```diff
@@ -125,7 +129,7 @@ DEFAULT_PGM_DIRECTION			EQU	1	; 1=Normal	2=Reversed	3=Bidir		4=Bidir rev
 DEFAULT_PGM_BEEP_STRENGTH		EQU	40	; 0..255 (BLHeli_S is 1..255)
 DEFAULT_PGM_BEACON_STRENGTH		EQU	80	; 0..255
 DEFAULT_PGM_BEACON_DELAY			EQU	4	; 1=1m		2=2m			3=5m			4=10m		5=Infinite
-DEFAULT_PGM_ENABLE_TEMP_PROT		EQU	7	; 0=Disabled	1=80C	2=90C	3=100C	4=110C	5=120C	6=130C	7=140C
+DEFAULT_PGM_ENABLE_TEMP_PROT		EQU	1	; 0=Disabled	1=80C	2=90C	3=100C	4=110C	5=120C	6=130C	7=140C
```
[the remark this answers is not in the document]
Comment (damosvil) [reply Bluejay.asm:128]: Fixed

File: Bluejay.asm [hunk 2]
```diff
@@ -107,7 +107,11 @@ Z_	EQU	26		; Bm Cm Am Vn __ RX __ __	Ac Ap Bc Bp Cc Cp __ __		yes	no	high	-	Pino
 ;PWM_FREQ			EQU	0	; 0=24, 1=48, 2=96 kHz
 
 
-PWM_CENTERED	EQU	DEADTIME > 0			; Use center aligned pwm on ESCs with dead time
+PWM_CENTERED					EQU		DEADTIME > 0			; Use center aligned pwm on ESCs with dead time
```
[the remark this answers is not in the document]
Comment (damosvil) [reply Bluejay.asm:110]: Fixed

File: Bluejay.asm [hunk 3]
```diff
@@ -223,16 +227,15 @@ Wt_Zc_Tout_Start_H:			DS	1	; Timer3 start point for zero cross scan timeout (hi
 Wt_Comm_Start_L:			DS	1	; Timer3 start point from zero cross to commutation (lo byte)
 Wt_Comm_Start_H:			DS	1	; Timer3 start point from zero cross to commutation (hi byte)
 
-Pwm_Limit:				DS	1	; Maximum allowed pwm (8-bit)
+Pwm_Limit:					DS	1	; Maximum allowed pwm (8-bit)
```
[the remark this answers is not in the document]
Comment (damosvil) [reply Bluejay.asm:226]: Fixed

File: Bluejay.asm [hunk 4]
```diff
@@ -223,16 +227,15 @@ Wt_Zc_Tout_Start_H:			DS	1	; Timer3 start point for zero cross scan timeout (hi
...
 Pwm_Limit_By_Rpm:			DS	1	; Maximum allowed pwm for low or high rpm (8-bit)
 Pwm_Limit_Beg:				DS	1	; Initial pwm limit (8-bit)
 
 Pwm_Braking_L:				DS	1	; Max Braking pwm (lo byte)
 Pwm_Braking_H:				DS	1	; Max Braking pwm (hi byte)
 
-Adc_Conversion_Cnt:			DS	1	; Adc conversion counter
-Current_Average_Temp:		DS	1	; Current average temperature (lo byte ADC reading, assuming hi byte is 1)
 Temp_Prot_Limit:			DS	1	; Temperature protection limit
+Temp_Pwm_Level_Setpoint:	DS	1	; PWM level setpoint
```
[the remark this answers is not in the document]
Comment (damosvil) [reply Bluejay.asm:234]: Fixed

File: Bluejay.asm [hunk 5]
```diff
@@ -431,6 +438,7 @@ LOCAL wait_for_t3 done_waiting
 
 wait_for_t3:
 	jnb	Flag_Timer3_Pending, done_waiting
+	Wdt_Kick_Enable						; Kick watchdog
```
[the remark this answers is not in the document]
Comment (damosvil) [reply Bluejay.asm]: WDT code removed in last commit

File: Bluejay.asm [hunk 6]
```diff
@@ -2447,8 +2549,8 @@ comp_read_wrong_extend_timeout:
 comp_read_wrong_timeout_set:
 	mov	TMR3CN0, #04h				; Timer3 enabled and interrupt flag cleared
 	setb	Flag_Timer3_Pending
-	orl	EIE1, #80h				; Enable Timer3 interrupts
-	jmp	comp_start				; If comparator output is not correct - go back and restart
+	orl	EIE1, #80h					; Enable Timer3 interrupts
```
[the remark this answers is not in the document]
Comment (damosvil) [reply Bluejay.asm:2545]: Fixed

File: Bluejay.asm [hunk 7]
```diff
@@ -2505,7 +2607,7 @@ setup_comm_wait:
 
 	setb	Flag_Timer3_Pending
 	orl	EIE1, #80h				; Enable Timer3 interrupts
-	setb	IE_EA					; Enable interrupts again
+	setb	IE_EA				; Enable interrupts again
```
[the remark this answers is not in the document]
Comment (damosvil) [reply Bluejay.asm:2603]: Fixed

File: Bluejay.asm [hunk 8]
```diff
@@ -3706,7 +3852,8 @@ IF MCU_48MHZ == 1
 	mov	SFRPAGE, #00h
 ENDIF
 	Initialize_Crossbar				; Initialize the crossbar and related functionality
-	call	switch_power_off			; Switch power off again, after initializing ports
+	call	switch_power_off		; Switch power off again, after initializing ports
```
[the remark this answers is not in the document]
Comment (damosvil) [reply Bluejay.asm:3848]: Fixed

File: Bluejay.asm [hunk 9]
```diff
@@ -3723,7 +3870,7 @@ ENDIF
 	clr	IE_EA					; Disable interrupts explicitly
 	call	wait100ms					; Wait a bit to avoid audible resets if not properly powered
 	call	play_beep_melody			; Play startup beep melody
-	call	led_control				; Set LEDs to programmed values
+	call	led_control			; Set LEDs to programmed values
```
[the remark this answers is not in the document]
Comment (damosvil) [reply Bluejay.asm:3865]: Fixed

File: Bluejay.asm [hunk 10]
```diff
@@ -3791,10 +3938,11 @@ setup_dshot:
 	mov	IT01CF, #(08h + (RTX_PIN SHL 4) + RTX_PIN) ; Route RCP input to Int0/1, with Int0 inverted
 
 	clr	Flag_Telemetry_Pending		; Clear DShot telemetry flag
+	clr	Flag_Ext_Tele				; Clear extended telemetry enabled flag
 
 	; Setup interrupts
 	mov	IE, #2Dh					; Enable Timer1/2 interrupts and Int0/1 interrupts
-	mov	EIE1, #80h				; Enable Timer3 interrupts
+	mov	EIE1, #80h					; Enable Timer3 interrupts
```
[the remark this answers is not in the document]
Comment (damosvil) [reply Bluejay.asm:3937]: Fixed

File: Bluejay.asm [hunk 11]
```diff
@@ -3959,26 +4108,16 @@ motor_start:
...
-
-	mov	Adc_Conversion_Cnt, #8		; Make sure a temp reading is done
-	call	check_temp_and_limit_power
-	mov	Adc_Conversion_Cnt, #8		; Make sure a temp reading is done next time
+	mov Ext_Telemetry_H, #0				; Clear extended telemetry data
+	mov	Temp_Pwm_Level_Setpoint, #255	; Initialize temperature pwm level setpoint
 
 	; Set up start operating conditions
-	clr	IE_EA					; Disable interrupts
+	clr	IE_EA						; Disable interrupts
```
[the remark this answers is not in the document]
Comment (damosvil) [reply Bluejay.asm:4107]: Fixed

File: Bluejay.asm [hunk 12]
```diff
@@ -4239,6 +4378,11 @@ exit_run_mode_on_timeout:
 exit_run_mode:
 	clr	IE_EA					; Disable all interrupts
 	call	switch_power_off
+
+	; Disable wdt and wait before posibly reenable shortly again
```
[the remark this answers is not in the document]
Comment (damosvil) [reply Bluejay.asm]: WDT code removed from last commit

File: Layouts/Base.inc [hunk 13]
```diff
@@ -315,8 +315,8 @@ $endif
 ;**** **** **** **** ****
 ; ADC and temperature measurement
 ;**** **** **** **** ****
-TEMP_LIMIT		EQU	49			; Temperature measurement ADC value for which main motor power is limited at 80degC (low byte, assuming high byte is 1)
-TEMP_LIMIT_STEP	EQU	9			; Temperature measurement ADC value increment for another 10degC
+TEMP_LIMIT		EQU	49				; Temperature measurement ADC value for which main motor power is limited at 80degC (low byte, assuming high byte is 1)
```
[the remark this answers is not in the document]
Comment (damosvil) [reply Layouts/Base.inc:318]: Fixed

File: Layouts/Base.inc [hunk 14]
```diff
@@ -315,8 +315,8 @@ $endif
...
+TEMP_LIMIT_STEP	EQU	9				; Temperature measurement ADC value increment for another 10degC
```
[the remark this answers is not in the document]
Comment (damosvil) [reply Layouts/Base.inc:319]: Fixed

File: Bluejay.asm [hunk 15]
```diff
@@ -170,19 +174,19 @@ Flag_Dir_Change_Brake		BIT	Flags1.5		; Set when braking before direction change
...
+;Flag_Clock_At_48MHz			BIT	Flags2.4		; Set if 48MHz MCUs run at 48MHz
 Flag_Rcp_Stop				BIT	Flags2.5		; Set if the RC pulse value is zero or if timeout occurs
 Flag_Rcp_Dir_Rev			BIT	Flags2.6		; RC pulse direction in bidirectional mode
 Flag_Rcp_DShot_Inverted		BIT	Flags2.7		; DShot RC pulse input is inverted (and supports telemetry)
 
 Flags3:					DS	1			; State flags. NOT reset upon motor_start
 Flag_Telemetry_Pending		BIT	Flags3.0		; DShot telemetry data packet is ready to be sent
 Flag_Dithering				BIT	Flags3.1		; PWM dithering enabled
-Flag_Had_Signal			BIT	Flags3.2		; Used to detect reset after having had a valid signal
+Flag_Had_Signal				BIT	Flags3.2		; Used to detect reset after having had a valid signal
```
[the remark this answers is not in the document]
Comment (damosvil) [reply Bluejay.asm:185]: Fixed

File: Bluejay.asm [hunk 16]
```diff
@@ -3959,26 +4108,16 @@ motor_start:
...
-
-	mov	Current_Average_Temp, ADC0L	; Read initial temperature
-	mov	A, ADC0H
-	jnz	($+5)					; Is reading below 256?
-	mov	Current_Average_Temp, #0		; Yes - set average temperature value to zero
-
-	mov	Adc_Conversion_Cnt, #8		; Make sure a temp reading is done
-	call	check_temp_and_limit_power
-	mov	Adc_Conversion_Cnt, #8		; Make sure a temp reading is done next time
+	mov Ext_Telemetry_H, #0				; Clear extended telemetry data
```
[the remark this answers is not in the document]
Comment (damosvil) [reply Bluejay.asm:4103]: Fixed

File: Common.inc [hunk 17]
```diff
@@ -181,7 +179,27 @@ Set_MCU_Clk_48MHz MACRO
 	mov	SFRPAGE, #00h
 
 	mov	CLKSEL, #03h				;; Set clock to 48MHz (Oscillator 1)
+ENDM
+
+Wdt_Configure MACRO
+IF USE_WATCHDOG == 1
+	mov LFO0CN, #83h				; Enable Low Frequency Oscillator and set to 80Khz in preparation to later enable WDT
+	mov WDTCN,#0 					; Set WDT interval to 0,8 ms
```
[the remark this answers is not in the document]
Comment (damosvil) [reply Common.inc]: WDT code removed in last commit

File: Common.inc [hunk 18]
```diff
@@ -181,7 +179,27 @@ Set_MCU_Clk_48MHz MACRO
...
+	nop 							; insert wait for 1 SYSCLK period
+ENDIF
+ENDM
 
-	setb	Flag_Clock_At_48MHz
+Wdt_Kick_Enable MACRO
+IF USE_WATCHDOG == 1
+	mov WDTCN,#0A5h					; reset watchdog timer
```
[the remark this answers is not in the document]
Comment (damosvil) [reply Common.inc]: WDT code removed in last commit

File: Common.inc [hunk 19]
```diff
@@ -181,7 +179,27 @@ Set_MCU_Clk_48MHz MACRO
...
+	; ensure insert wait for 2 divided LFOSC0 clock periods before kicking again, otherwise wdg reset is canceled
+ENDIF
 ENDM
 
+Wdt_Disable MACRO
+IF USE_WATCHDOG == 1
+	mov WDTCN,#0DEh					; disable software watchdog timer
```
[the remark this answers is not in the document]
Comment (damosvil) [reply Common.inc]: WDT code removed in last commit

File: Bluejay.asm [hunk 20]
```diff
@@ -170,19 +174,19 @@ Flag_Dir_Change_Brake		BIT	Flags1.5		; Set when braking before direction change
...
 
 Flags2:					DS	1			; State flags. NOT reset upon motor_start
-;						BIT	Flags2.0
+Flag_Ext_Tele				BIT	Flags2.0		; Set if extended telemetry is enabled
 Flag_Pgm_Dir_Rev			BIT	Flags2.1		; Set if the programmed direction is reversed
 Flag_Pgm_Bidir				BIT	Flags2.2		; Set if the programmed control mode is bidirectional operation
-Flag_Skip_Timer2_Int		BIT	Flags2.3		; Set for 48MHz MCUs when Timer2 interrupt shall be ignored
-Flag_Clock_At_48MHz			BIT	Flags2.4		; Set if 48MHz MCUs run at 48MHz
+Flag_32ms_Elapsed			BIT	Flags2.3		; Set when timer2 interrupt is triggered
+;Flag_Clock_At_48MHz			BIT	Flags2.4		; Set if 48MHz MCUs run at 48MHz
```
[the remark this answers is not in the document]
Comment (damosvil) [reply Bluejay.asm:177]: No longer used

**Comment by stylesuxx** [comment]
**ATTENTION**: This PR continues in #20 

There were some issues with the PR, switching from old repo to detached repo and thus @damosvil made a new PR (with feedback from here - at least partially - applied). I'll leave this open since it has been referenced in other issues and close it once we finish the NEW PR #20

**Comment by stylesuxx** [comment]
Closing this since superseding #20 is merged.

**Pull request closed without merging**